### ISSUE TITLE
Start Interval Tasks at Startup

### ISSUE THREAD

**Issue by DCPDamon** [M-Files/VAF.Extensions.Community#34]
Is it possible to have Tasks based on Interval to be triggered at startup?  Right now it's queueing the task as expected, but it would be nice to run initially too.  Probably, best long term solution would allow developer to hard code in an Attribute whether it should run at startup - or someone make that option visible in Configuration for admins to change at runtime if needed.

**Comment by CraigHawker** [comment]
As there are no attributes on the method that define the interval (it's done by exposing configuration, and hooking that to the processor), my initial gut instinct is that this could be part of `RecurringOperationConfigurationAttribute:`

```
[RecurringOperationConfiguration(VaultApplication.QueueId, VaultApplication.ImportDataFromRemoteSystemTaskType, RunAtStartup = true] // Default for RunAtStartup is false.
```

To make it configurable then perhaps it could be part of `Frequency`.  This is the only class we have that might give us the option to add a configuration item, although it then allows the user to also migrate to a schedule-based configuration if they wanted.

Any thoughts/comments on that implementation?

**Comment by DCPDamon** [comment]
Having it as an attribute of 'RecurringOperationConfiguration' is definitely a great place for it; and desired functionality.  Personally, I'm using Frequency for all my configs to give admins maximum flexibility.  I would not be opposed to it being exposed as config.  My question would be, would you have it as an attribute and config? If so how would they interact? Or is this a scenario where we can only have 1 option.  If we can only have 1, probably go with attribute, so that someone defining their config as TimeSpan instead of Frequency can define RunAtStartup.

**Comment by CraigHawker** [comment]
If we had both available then my gut - again: feel free to argue - is that the hard-coded value on the attribute would be used only if the Frequency was not configured.  Were the frequency populated it would override the hard-coded value.

**Comment by DCPDamon** [comment]
Agreed! But what would it looks like in Admin tool? If it's "(not set)" in admin tool then use the attribute? Ideally Admin would tell the user what the default is (based on how the attribute is configured.

**Comment by CraigHawker** [comment]
It would not be possible to automate that; the developer would have to provide a reasonable textual description in the DefaultValue property of the attribute, i.e.:

```
[RecurringOperationConfiguration
(
    VaultApplication.QueueId,
    VaultApplication.ImportDataFromRemoteSystemTaskType,
    RunAtStartup = true,
    DefaultValue = "Runs at vault startup, then every 10 minutes"
)]
public Frequency Schedule { get; set; } = new TimeSpan(0, 10, 0);
```
or:
```
[RecurringOperationConfiguration
(
    VaultApplication.QueueId,
    VaultApplication.ImportDataFromRemoteSystemTaskType,
    RunAtStartup = false,
    DefaultValue = "Runs every 10 minutes, starting from the time the vault starts"
)]
public Frequency Schedule { get; set; } = new TimeSpan(0, 10, 0);
```

**Comment by DCPDamon** [comment]
That works for me it it works for you.

**Comment by CraigHawker** [comment]
I'm comfortable with this as an approach.  I personally don't have a lot of time over the next couple of weeks, but I think it'd be great to try and get this in the public extensions release that targets the VAF 2.3.  I'm going to add the "help wanted" label but I expect that I'll be the one doing it, if time permits.

**Comment by CraigHawker** [comment]
Upon investigating this, it does already do this for interval-based configuration.  The reason for this was to be broadly in-line with how the older Background Operations ran (as they worked this way).  You can see this in the [`RecurringOperationConfigurationManager`](https://github.com/M-Files/VAF.Extensions.Community/blob/vaf2.3-compatibility/MFiles.VAF.Extensions/RecurringOperationConfigurationManager.cs#L151) when it populates the recurring operation configuration from the loaded configuration.

However, it does this by specifically looking for `TimeSpan`-based properties and fields (i.e. it *only* works for intervals).  It will not behave this way for `Schedule`- or `Frequency`-based configuration items.

Can we go back and just confirm that this is the behaviour that you're seeing, and that there's not some other bug somewhere?

**Comment by DCPDamon** [comment]
OK, I confirmed in a test that `TimeSpan` tasks will run at startup.  Just not `Frequency` (I didn't test `Schedule`).

**Comment by CraigHawker** [comment]
An initial implementation is in branch https://github.com/M-Files/VAF.Extensions.Community/tree/issue-34/MFiles.VAF.Extensions.

Specifically:
* Developers can expose their interval as a [`TimeSpanEx`](https://github.com/M-Files/VAF.Extensions.Community/blob/issue-34/MFiles.VAF.Extensions/Configuration/TimeSpanEx.cs) (instead of `TimeSpan`) which allows the user to toggle off the auto-tun-on-vault-start: https://github.com/M-Files/VAF.Extensions.Community/tree/issue-34/MFiles.VAF.Extensions#suppressing-an-interval-based-task-from-running-at-vault-startup
  *  Small note here is that the current implementation doesn't allow an old TimeSpan configuration to be deserialised into a TimeSpanEx, so beware of just changing the type as it won't load existing configuration properly.
* The [`Schedule`](https://github.com/M-Files/VAF.Extensions.Community/blob/issue-34/MFiles.VAF.Extensions/Configuration/ScheduledExecution/Schedule.cs) now has a property, which defaults to false, for whether to run at startup.
* The [`Frequency`](https://github.com/M-Files/VAF.Extensions.Community/blob/issue-34/MFiles.VAF.Extensions/Configuration/Frequency.cs) exposes a similar property, but delegates to the selected configuration.
* A flag is passed into the [RecurringOperationConfigurationManager](https://github.com/M-Files/VAF.Extensions.Community/blob/issue-34/MFiles.VAF.Extensions/RecurringOperationConfigurationManager.cs#L151) to show whether it's the vault start or a re-configuration event.  This is used then to calculate whether to run the task "now" or to wait until the next elapsed/configured time.

Interested in feedback/comments on this.

**Comment by CraigHawker** [comment]
The fact that moving from TimeSpan to TimeSpanEx would break an existing configuration concerned me, so I have a unit-tested approach for that now in the branch: https://github.com/M-Files/VAF.Extensions.Community/commit/8f34b1d4b2d4ee4cea38f3566d21dadb1ded1337

This needs to go through more of a formal review before merging.  I'll see whether I can find someone to volunteer for that over the next few days.

**Comment by CraigHawker** [comment]
Changes have been merged into the VAF 2.3 branch and will be included in the next preview.
